beetsplug/acoustic.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+# This file is part of beets.
+# Copyright 2016, Ohm Patel.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+""" Fetch various AcousticBrainz metadata using MBID
+"""
+from __future__ import (division, absolute_import, print_function,
+                        unicode_literals)
+
+import requests
+
+from beets import plugins, ui
+
+ACOUSTIC_URL = "http://acousticbrainz.org/"
+LEVEL = "/high-level"
+
+
+class AcousticPlugin(plugins.BeetsPlugin):
+    def __init__(self):
+        super(AcousticPlugin, self).__init__()
+
+    def commands(self):
+        cmd = ui.Subcommand('acoustic',
+                            help="fetch metadata from AcousticBrainz")
+
+        def func(lib, opts, args):
+            fetch_info(lib)
+
+        cmd.func = func
+        return [cmd]
+
+
+def fetch_info(lib):
+    """Currently outputs MBID and corresponding request status code
+    """
+    for item in lib.items():
+        if item.mb_trackid:
+            rs = requests.get(generate_url(item.mb_trackid)).json()
+
+            item.abrainz_dance = get_value(rs, ["highlevel",
+                                                "danceability",
+                                                "all",
+                                                "danceable"])
+            item.abrainz_happy = get_value(rs, ["highlevel",
+                                                "mood_happy",
+                                                "all",
+                                                "happy"])
+            item.abrainz_party = get_value(rs, ["highlevel",
+                                                "mood_party",
+                                                "all",
+                                                "party"])
+
+            item.write()
+            item.store()
+
+
+def generate_url(mbid):
+    """Generates url of AcousticBrainz end point for given MBID
+    """
+    return ACOUSTIC_URL + mbid + LEVEL
+
+
+def get_value(data, map_path):
+    """Allows traversal of dictionary with cleaner formatting
+    """
+    return reduce(lambda d, k: d[k], map_path, data)
```

Comment:
Clever!

Let's use some error handling here too. In case we're wrong about the format of the JSON, this will raise a `KeyError`. The plugin should avoid crashing, but it's up to you exactly what it does. Options include:

* Print an error and `continue` onto the next track.
* Print an error and stop processing.
* Just don't assign that particular value.